Stripes detector and a mask generator

dkazanc opened the pull request:
Frequently we've got data with stripes artifacts in it which are needed to be removed.
![sino](https://user-images.githubusercontent.com/6651977/219389133-2f7fd945-55ad-49a0-9b89-fd76d4cd86f6.png)

 The traditional approach would be to apply [stripes/ring removal algorithms](https://tomopy.readthedocs.io/en/stable/api/tomopy.prep.stripe.html#). This is, however, a bit feature-unaware approach in which stripe-alike data features can be interpreted as stripes and therefore removed or distorted leading to problems in the reconstructed data.

In this PR, we offer a functionality that can highlight various types of vertical linear features (a stripes detection step) and another module which can process those weights and generate a binary mask for stripes only by excluding useful data features. This mask can then be forwarded to another module to inpaint the missing data (we will create a separate PR on that) or a stripe removal algorithm. 

Therefore in this PR:
- **stripes detecton method.** This is a simple approach which relies on the fact that in the gradient space the ratio of pixels in the direction orthogonal to the vertical feature is normally different from the pixels that are parallel to the feature. We calculate a weight map and process it with a vertical 1D median filter which further highlights the stripes present. As a result of this method we will have a [0,1] range weight map which can be thresholded. The smaller weights represent stripes. The approach can also mistakenly identify the features that belong to the data and should not be removed, therefore the second step could be helpful. 
![tomopy0002](https://user-images.githubusercontent.com/6651977/219391607-f36f707c-9486-43b0-94cf-a512d666de3a.jpg)

- **binary mask generation method** This method takes the weights obtained with the previous detection step and performs a number of tests to include or exclude a certain weight from the mask. The method relies on user provided parameters that define the nature of stripes present in the data, e.g. length, width, depth. This is a vital step to achieve a good binary mask which contain only stripes, hence more parameters to control. However, we believe that those parameters can remain fixed for the same dimensions of the data. 
![tomopy0000](https://user-images.githubusercontent.com/6651977/219392494-56cb1e51-ffa8-42f4-b627-9ead38bcd805.jpg)

The proposed methods can be helpful in the circumstances when a special cautious approach needed to stripes detection and removal. Also it can be helpful in a situation when stripes are very sharp and the data is not possible to recover in those regions, then the inpainting approach usually needed. The inpaintng relies on a mask provided.  We will later provide inpainting methods that can fill the missing data by using the masks generated in this PR.

carterbox commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,156 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+    this orientation, the stripes are the vertical features. The method works with
+    full and partial stripes of constant ot varying intensity. 
+
+    .. versionadded:: 1.13
```
```suggestion
    .. versionadded:: 1.14
```
We use semantic versioning. Since 1.13 is already released, this feature will first appear in 1.14.

dkazanc replied:
fixed in 4974eee

carterbox commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,156 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
         tomo[:, m, :] = sino
+
+
+def stripes_detect3d(arr, vert_filter_size_perc=5, radius_size=3, ncore=None):
```
Please change the `arr` argument so it is consistent with the other stripe methods. i.e. It should be named `tomo` and have the same axis ordering as the other methods in this module (I'm actually not sure what that is because the docs aren't great).

dkazanc replied:
fixed in 4974eee 
I renamed input `arr` to `tomo` for one function and to `weights` for another. Also I checked the conventional axis ordering for `tomo` and it seems to be consistent with` [angles, vertical_detector, horizontal_detector]` .

carterbox commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,156 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+    else:
+        raise ValueError("The input array must be a 3D array")
+
+    # calculate absolute values based on the provided percentages: 
+    if 0.0 < vert_filter_size_perc <= 100.0:
+        vertical_filter_size = (int)((0.01*vert_filter_size_perc)*dz)
+    else:
+        raise ValueError("vert_filter_size_perc value must be in (0, 100] percentage range ")
```
Why is the Python API so different from the C API? Why not just expose `vertical_filter_size` directly? Accepting a float value that is cast down to an integer gives the illusion of greater precision?

dkazanc replied:
very good point. I've been thinking about this one. The reason why I went for more relative to the sizes of the data parameter is to avoid a problem of redefining this parameter every time when the sizes of data change. Also you need to know the sizes of the data to set it to something meaningful, while in precents you can  take something approximate and will work. 
I'm not 100% about this one, would you rather go for an absolute value?

dkazanc replied:
They are independent indeed, but the stripe's width can be relative to the total size of the detectors array. I guess for me it was easier to relate to the total size, e.g., 1% of the total detector size rather than to pick an exact absolute value in pixels. This choice is obviously subjective and I'm happy to change to absolute values. Obviously for absolute values you would certainly need to change them if the data dimensions change, they are not scalable.

dkazanc replied:
thought more about it and decided to revert to absolute values. Will commit the changes soon.

dkazanc commented on include/libtomo/stripes_detect3d.h:
```diff
@@ -0,0 +1,68 @@
+                         int stripe_depth_min,
+                         int stripe_width_min,
+                         float sensitivity,
+                         int ncores, int dimX, int dimY, int dimZ);
```
fixed in 4974eee

dkazanc commented on source/libtomo/misc/utils.c:
```diff
@@ -200,3 +200,55 @@ quicksort_uint16(unsigned short* x, int first, int last)
         quicksort_uint16(x, j + 1, last);
     }
 }
+
```
partially fixed in 4974eee. Should we rename _utils.c_ to something more meaningful and move to a different location? Accessible for different methods. Although I moved the gradient3D function I still need things like sorting from utils. These are things which are frequently required.

dkazanc replied:
Thanks, possibly splitting _utils_ into separate modules. For instance, sorting.c and gradient.c makes it more explicit what they do?

carterbox replied:
What do you think about removing the sorting implementations completely? I noticed that bubble sort implementations are unused, and we could use [quicksort from the C standard library](https://www.tutorialspoint.com/c_standard_library/c_function_qsort.htm).

Then the CopyIm functions could be in their own header-only module (they are very short).  I just need to figure out the proper way of sharing a non-public header-only module between CMake modules.

carterbox replied:
Actually CopyIm could also be replaced with [`memcpy`](https://www.tutorialspoint.com/c_standard_library/c_function_memcpy.htm) from the C standard library. Then this module is no longer needed.

dkazanc replied:
thanks, I'll have a look reorganising this!

nghia-vo commented:
There's a method to detect stripe available in Tomopy: https://github.com/tomopy/tomopy/blob/14f5d80624a7540f3515388492570de6e3a1f8ba/source/tomopy/prep/stripe.py#L680. From a sinogram, there's a pre-processing step :  https://github.com/tomopy/tomopy/blob/14f5d80624a7540f3515388492570de6e3a1f8ba/source/tomopy/prep/stripe.py#L716
to get a 1d-array then pass it to the detection method. We just need to modify these functions to return a mask. Details of the method are here: https://sarepy.readthedocs.io/toc/section3_1/section3_1_3.html. Is there an advantage of the proposal approach? There're a list of sinograms here to compare: https://github.com/nghia-vo/sarepy/tree/master/data

nghia-vo commented:
About the concern of removing stripe-alike data features, remove_all_stripe and remove_stripe_based_sorting can deal well with that as demonstrated here: https://algotom.readthedocs.io/en/latest/toc/section4/section4_4.html#having-valid-stripes-not-artifacts

dkazanc commented:
reply to @nghia-vo. 
There are several key differences to the [method](https://github.com/tomopy/tomopy/blob/14f5d80624a7540f3515388492570de6e3a1f8ba/source/tomopy/prep/stripe.py#L680)  that is available in TomoPy for stripe detection.
- It is a fully 3D method compare to 1D approach in TomoPy. The method takes the advantage of 3d spatial information for detection and also for mask generation where useful features can be excluded. 
- The method relies on a completely different methodology and can detect of stripes of any length - full or partial, also as it shows in practice of varying intensity. The 1d method will get false alarms for partial stripes. 
- The method specifically consists of two parts where mask generation is another algorithm to better control the stripes detection process.
- Finally it is implemented in C and takes the opportunity of performing quite time-consuming 3D detection process reasonably fast. 

There is no claim whatsoever for this method to be a superior or a general solution for all problems. My student and I have been working on a DL project where accurate [detection](https://github.com/dkazanc/NoStripesNet/blob/real_data/utils/stripe_detection.py) is required. We can also consider adding other algorithms for detection like cluster-based detection using GMM for instance. 

Now playing with the [method](https://github.com/tomopy/tomopy/blob/14f5d80624a7540f3515388492570de6e3a1f8ba/source/tomopy/prep/stripe.py#L680)  in TomoPy and varying parameters I couldn't get the less visible stripe detected as shown above. You either get lots of false alarms (left) or just one stripe (right). 
![image](https://user-images.githubusercontent.com/6651977/220200134-74b38295-6612-4ad1-bef2-1e482eac575b.png)

dkazanc commented on cmake/Modules/Packages.cmake:
```diff
@@ -219,12 +219,12 @@ if(TOMOPY_USE_CUDA)
     # Pascal support 70, 72      + Volta support 75          + Turing support
     if(NOT DEFINED CUDA_ARCH)
       if(CUDAToolkit_VERSION_MAJOR VERSION_LESS 11)
-        set(CUDA_ARCH "30;32;35;37;50;52;53;60;61;62;70;72;75")
```
thanks, that would be quite helpful as my build simply fails as the lower than 50 architectures being obsolete apparently. I was thinking that may be another way remove them is to look for CUDA SDK version available, but your way seems better.

carterbox replied:
Resolved in #606

nghia-vo commented:
@dkazanc : The question is not about technique. It's about what benefit it can bring to users. From method development point of view, we should compare the performance of the proposal method on various types of sinograms to see the benefits. From software development point-of-view, the code base of this PR is pretty big, it will be difficult to maintain if you or Tomopy's developers with C experience no longer work on the project. 
Back to your comparison with the current detection method, the final aim of a removal method is to remove stripes not detect stripes. Detecting stripe is just a step in some of methods in Tomopy. For other methods, they remove stripes by equalizing the response of neighbouring pixels. For users' benefit, we should compare the removal effect instead. 
Certainly, for you and your student project, detecting stripe is more important but for general users they may care more about removing stripes.

dkazanc commented:
@nghia-vo 
- _compare the performance of the proposal method on various types of sinograms to see the benefits_. 
We already compared the performance on 3d datasets. The method needs full 3d projection data, not sinograms alone. We can share the data or if there is data somewhere we're happy to compare. 
- _From software development point-of-view, the code base of this PR is pretty big, it will be difficult to maintain if you or Tomopy's developers with C experience no longer work on the project_. 
I don't see any issue with that as the code base of TomoPy is filled with C functions of [various complexity](https://github.com/tomopy/tomopy/tree/master/source/libtomo). I believe as this is partially a C-based project, there will be developers around to fix and maintain it?
- _the final aim of a removal method is to remove stripes not detect stripes._
Have to disagree here again as I think that the stripes detection problem is crucial to any removal algorithms and we can provide many examples when the stripes removal algorithms make things worse by distorting the data. The detection is as important and requires special attention. 
- _For users' benefit, we should compare the removal effect instead._
We will add the inpainting methods that operate with the provided masks when this PR is complete. This is not purely stripe-related but a general feature. We believe that is can be useful to the community.

nghia-vo commented:
Sound good. This dataset here: https://doi.org/10.5281/zenodo.1443568 has all types of stripes in it and can be useful for testing both detection and removal method.

dkazanc commented:
rigorously speaking the methods should be tested with simulated data and quantified by some metric. But OK we can try this, although it is subjective.

dkazanc commented:
- Fixed the data type overflow issue for big volumes. The casting to indices wasn't correct which was leading to segfault on very large volumes.
- Reverting to absolute parameters instead of the ones in precents
- Some improvements to the mask generator method, improved stability and accuracy. 
- Tested on the full data of size (1801, 2160, 2560). The total time of computation is 2 hours on 40 cores. The algorithm can be considerably accelerated with GPUs.

dkazanc commented:
@nghia-vo 
I've tried the algorithm on the raw data you provided. Here are the normalised [sinograms](https://dlsltd-my.sharepoint.com/:u:/g/personal/daniil_kazantsev_diamond_ac_uk/EfqmXxbgtmdInfe27kf_534Bds_jW2wdFZqeCuqZg6anAA?e=zA1OXP) and the associated [masks](https://dlsltd-my.sharepoint.com/:u:/g/personal/daniil_kazantsev_diamond_ac_uk/EcB5j331D0FMl8SP4qPmeg4BcDXdfXyE-sQdxAzEiDk5sA?e=iLA9sJ). Overall I think quite satisfactory performance, although there are some misclassification and some false alarms. However, some nice detections on slices 685 and 1369  for example. At 1369 (the second image) the elongated feature that belongs to a sample was left alone, this is pretty difficult to classify if the method is 1d or 2d. 
Some examples: 
![Screenshot from 2023-02-25 18-54-51](https://user-images.githubusercontent.com/6651977/221552009-b6b6d225-f7cb-48a2-a30c-e437a5d7203a.png)
![Screenshot from 2023-02-26 12-34-22](https://user-images.githubusercontent.com/6651977/221552021-ad598980-0d0d-441d-8f1a-21762305a3fa.png)
![Screenshot from 2023-02-27 11-32-15](https://user-images.githubusercontent.com/6651977/221553179-cfa63255-7d68-4188-b32b-cd794be58f5e.png)

nghia-vo commented:
Thanks Daniil, there're practical aspects you may want to consider:
- Due to the scattering effect of scintillation photon, the occurrence of stripes is [sample-dependent](https://sarepy.readthedocs.io/toc/section1.html#id6). This makes a generated mask for one sample can't simply be applied to other samples of the same experiment setup. 
- If the method relies on a median filter, [blurry stripes](https://sarepy.readthedocs.io/toc/section2.html#id6) will be hard to detect.
- Look like the method requires large RAM to load 3d array, maybe it can be implemented chunk-by-chunk to allow users with small RAM computer can try it.

dkazanc commented:
thanks for enabling `memcpy` @carterbox. I've replaced sorting with the `qsort` function and removed utils. When doing that I found a bug with the median wrapper for non contiguous arrays being passed to C function, so I fixed that as well. The tests run, the processing seems working, so I'm happy for merging. cheers

carterbox commented on include/libtomo/stripe.h:
```diff
@@ -53,3 +53,21 @@
+                           int window_halflength_vertical,
+                           int ratio_radius,
+                           int ncores,
+                           long dimX, long dimY, long dimZ);
```
Please use `int` for the array shapes. Internally, you can convert to `size_t` if you need to compute the number of elements in the array. The maximum integer on any platform should be sufficient to express any single dimension of any array.

dkazanc commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,151 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+    Raises
+    ------
+    ValueError
+        If the input array is not three dimensional.        
```
I've removed the strict [0,1] condition as it is only a preferable one. Practically, some values in -log(normalised data) could be larger than one and also negative, this shouldn't be a problem for the method. I've added this in the docs.

dkazanc commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,151 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+        3D weights array, a result of stripes_detect3d module given in
+        [angles, detY(depth), detX] axis orientation.
+    threshold : float, optional
+        Threshold for the given weights, the smaller values correspond to the stripes
```
Thanks, fair point. I've re-written the description to: _This parameter defines what weights will be considered as potential candidates for stripes. It is important to remove as many false outliers as possible by taking the highest acceptable value. The lower value might lead to ignore the actual stripes and the higher one can generate false alarms._

Hopes it is clearer now.

carterbox replied:
This is a 'less than' threshold or 'greater than' threshold? These explanations are contradictory because they claim raising the threshold both increases the number of false alarms and removes outliars?

Are lower valued weights more likely or less likely to be stripes?

dkazanc replied:
I understand how a bit confusing the description might sound.

_Are lower valued weights more likely or less likely to be stripes?_ The lower weights are **more** likely to be stripes.

 Basically if we take the threshold level too low, e.g., lower than 0.5 normally, then we will get the most prominent stripes only and the weak stripes will be mostly ignored. It is tricky to find the exactly right threshold value that would lead to both weak and prominent stripes detected equally well. So I'm suggesting taking the value when the false alarms just start to appear, hence "highest acceptable"...

dkazanc replied:
I slightly re-wrote that bit.

dkazanc commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,151 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+        [angles, detY(depth), detX] axis orientation.
+    threshold : float, optional
+        Threshold for the given weights, the smaller values correspond to the stripes
+    min_stripe_length : int, optional
+        Minimum accepted length of a stripe in pixels. Can be large if there are full stripes in the data.
+    min_stripe_depth : int, optional
+        Minimum accepted depth of a stripe in pixels. The stripes do not extend very deep, with this parameter more non-stripe features can be removed. 
+    min_stripe_width : int, optional
```
Although I understand the logic of why they should have the axis title in the name it can be that `length`, `depth` and `width` are more the stripe-related descriptors rather than the axis related? I've put the axis names in the description, let me know if it works for you?

carterbox replied:
This was helpful, but there is still double naming of axes which are confusing.

For example, "The pixel size of the vertical (angle dimension) " I think this statement means to imply that the angle dimension of the `tomo` array is also the vertical dimension, but ND-arrays don't have a vertical dimension, they have a 0th, 1st, 2nd, 3rd, ... dimensions.

Instead of "With this orientation, the stripes are the vertical features.", we should say "With this orientation, the stripes are features in axis 0" or "With this orientation, the stripes are features along the angle axis."

dkazanc replied:
OK, I think I understand your point here. I guess we're talking about two things - the algorithmic part and the data I/O part. I'm happy to go along with more abstract naming for dimensions and, as you suggest, refer to the data dimensions angle, detY, detX and not their spatial orientations with respect to data. Would you mind changing this or I can do it? thanks

dkazanc replied:
also corrected.

dkazanc commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,151 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+    input_type = weights.dtype
+    if (input_type != 'float32'):
+        weights = dtype.as_float32(weights)  # silent convertion to float32 data type
+    out = np.uint8(np.empty_like(weights, order='C'))
```
I guess you could have convert it to boolean on the wrapper side but sometimes it is easier to do some [morphological processing](https://scikit-image.org/docs/stable/api/skimage.morphology.html) with the mask obtained.

carterbox replied:
Do you mean some morphological processes take non-binary inputs? But this algorithm always returns a binary output?

dkazanc replied:
yes, it does always return the binary output. My point here was that for some morphological processing we can feed the binary data output [as is](https://scikit-image.org/docs/stable/auto_examples/applications/plot_morphology.html). I also would like to add a method which takes the binary mask to do the inpainting of the missing data region. So we will need convert boolean back and forth potentially?

carterbox replied:
I'm still confused. The purpose of this thread is to discuss why the current implementation uses `uint8` to store/represent a `bool` value. From my perspective, this makes the data eight times larger than it needs to be. For what reasons did you choose to use `uint8` instead of the `bool` datatype?

You linked the scipy morphology module which contains algorithms which can work with BOTH `bool` AND `uint8`, so I'm not sure what the benefit of using 8 bits to represent a bool mask.

dkazanc replied:
OK, I guess I decided not to deal with the conversion to `bool` in the wrapper as C doesn't  have a support for it. Memory-wise we would still need to create `uint8` as an output first before converting it to boolean in the wrapper. If you think we should store it as a boolean - I'm totally fine with it. Shall I implement this?

carterbox replied:
I am asking why the both python and C implementation uses uint8 instead of bool. We want to avoid conversions when possible.

dkazanc replied:
ok, I will rewrite to use `bool` instead `uint8` all the way.

carterbox commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,163 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+    """
+    Detect stripes in a 3D array. Usually applied to normalized projection data.
+
+    The method works with full and partial stripes of constant ot varying intensity. 
```
```suggestion
    The method works with full and partial stripes of constant of varying intensity. 
```

carterbox reviewed (changes requested):
Thank you for making my requested changes, I have two final questions about how to use the Python API. If you could answer then in the comments, then I'll fix up the docstrings.

dkazanc commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,211 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+        Minimum width of a stripe in pixels with respect to the "detX" axis.
+        The stripes that close to each other can be merged together with this
+        parameter.
+    sensitivity_perc : float, optional
+        The value in percents to impose less strict conditions on length, depth
+        and width parameters of a stripe.
```
Yes, 100% means that you're 100% strict, i.e., the values for stripe parameters will be preserved exactly as given - no deviation.

dkazanc commented on source/tomopy/prep/stripe.py:
```diff
@@ -939,3 +941,211 @@ def _remove_stripe_based_interpolation(tomo, snr, size, drop_ratio, norm):
+
+    Parameters
+    ----------
+    tomo : ndarray
+        3D tomographic data of float32 data type, preferably in the [0, 1]
+        range, although reasonable deviations accepted (e.g. the result of the
+        normalization and the negative log taken of the raw data). The
+        projection data should be given with [angles, detY(depth),
+        detX(horizontal)] axis orientation. With this orientation, the stripes
+        are the vertical features.
```
correct, the angles dimension corresponds to "vertical" orientation for this shape of the input data.
